README.md

```diff
@@ -24,90 +24,92 @@ The OpenCL *Installable Client Driver* extension (`cl_khr_icd`) is described in
 
-The OpenCL ICD Loader requires OpenCL Headers.
-To use system OpenCL Headers, please specify the OpenCL Header location using the CMake variable `OPENCL_ICD_LOADER_HEADERS_DIR`.
-By default, the OpenCL ICD Loader will look for OpenCL Headers in the `inc` directory.
+### Dependencies
 
-The OpenCL ICD Loader uses CMake for its build system.
+The OpenCL ICD Loader requires:
+- the [OpenCL Headers](https://github.com/KhronosGroup/OpenCL-Headers/).
+- The OpenCL ICD Loader uses CMake for its build system.
 If CMake is not provided by your build system or OS package manager, please consult the [CMake website](https://cmake.org).
-
-The Windows OpenCL ICD Loader requires the Windows SDK to check for and enumerate the OpenCLOn12 ICD.
+- The Windows OpenCL ICD Loader requires the Windows SDK to check for and enumerate the OpenCLOn12 ICD.
 An OpenCL ICD Loader can be built without a dependency on the Windows SDK using the CMake variable `OPENCL_ICD_LOADER_DISABLE_OPENCLON12`.
 This variable should only be used when building an import lib to link with, and must be enabled when building an OpenCL ICD Loader for distribution!
 
-### Build and Install Directories
-
-A common convention is to place the `build` directory in the top directory of the repository and to place the `install` directory as a child of the `build` directory.
-The remainder of these instructions follow this convention, although you may place these directories in any location.
-
-### Example Usage
+### Example Build
 
 For most Windows and Linux usages, the following steps are sufficient to build the OpenCL ICD Loader:
 
-1. Clone this repo:
+1. Clone this repo and the OpenCL Headers:
 
         git clone https://github.com/KhronosGroup/OpenCL-ICD-Loader
+        git clone https://github.com/KhronosGroup/OpenCL-Headers
 
-1. Obtain the OpenCL Headers, if you are not planning to use system OpenCL headers.
-Headers may be obtained from the [Khronos OpenCL Headers](https://github.com/KhronosGroup/OpenCL-Headers) repository.
+1. Install OpenCL Headers CMake package
 
-1. Create a `build` directory:
+        cmake -D CMAKE_INSTALL_PREFIX=./OpenCL-Headers/install -S ./OpenCL-Headers -B ./OpenCL-Headers/build 
+        cmake --build ./OpenCL-Headers/build --target install
 
-        cd OpenCL-ICD-Loader
-        mkdir build
-        cd build
+1. Build and install OpenCL ICD Loader CMake package. _(Note that `CMAKE_PREFIX_PATH` need to be an absolute path. Update as needed.)_
 
-1. Invoke `cmake` to generate solution files, Makefiles, or files for other build systems.
+        cmake -D CMAKE_PREFIX_PATH=/absolute/path/to/OpenCL-Headers/install -D CMAKE_INSTALL_PREFIX=./OpenCL-ICD-Loader/install -S ./OpenCL-ICD-Loader -B ./OpenCL-ICD-Loader/build 
```

Comment:
Perhaps you'd want to use CMake's `FetchContent` functionality here? We've recently switched several DirectX CMake projects to using it.

You can use `FetchContent_Declare` to provide the URL to the headers GitHub repo. If the target is not already present, you can use `FetchContent_MakeAvailable` to automatically download it into the build directory and include it.

If you want a parent project's `CMakeLists.txt` to be able to override it, then that parent build script can use a different `FetchContent_Declare`, which can point to a different repo (e.g. for internal builds that shouldn't reach out to the public internet), or can point to an already-cloned version of it (e.g. as a submodule of a parent project).

Reply:
My expectation is that once the SDK build definitions are updated and installing the SDK means "installing the SDK", then standalone builds of the ICD-Loader will become niche. If anything, only developers of runtimes will interact with this repo directly, and even they can tweak the loader inside the SDK. (That's only more involved it being a submodule.) OpenCL is in a state where LLVM was before going to a monorepo. It's insanely hard to deliver streamlined experience for both consumers of the monorepo and to those wanting to build the components in isolation. (With CMake that is.) The fact that it's possible is a feat already. The indication at the beginning of the build guide does suggest that building the SDK is simpler and OpenCL devs will likely want to interact with that. (With time they'll get Layers too along with the SDK.)

1. There are multiple solutions to this, if this is deemed to be painful.
1.1. Making the headers a submodule. This introduces diamond inheritance in the SDK, where both the ICD-Loader and CLHPP depend on the headers. Keeping the modules in sync is solvable, but given how we aleady have two versions of CMock in the SDK when recursively cloned... it can bite.
1.2. Looking for the target like you said and when not found using `add_external_project` which will make it part of the build. One of our libraries rocPRIM does this for eg. [here](https://github.com/ROCmSoftwarePlatform/rocPRIM/blob/develop/cmake/Dependencies.cmake#L108-L144). (It uses downloadProject which elevates configuring-building-installing to configuration time, but one loses multi-config generator support. I would use `add_external_project` as-is.) This approach comes at the cost of complexity, both for us and end-users, but the default we select would be more streamlined. This is once again borderline dep management territory.
2. I wrote the note because it did surprise me while writing the docs, but I've been developing this CMake rework for a year now (with pauses) and this is the first time I met this limitation. When I configured my IDE for SDK development as independent packages (not the recursive clone of the SDK repo), in my IDE's `settings.json` there is no inherent working directory and writing absolute paths was natural.

We can move in any direction. This part can be polished if we want to, but I don't expect users to be interacting diretly with this repo going forward. This is the "unfortunate" consequence of promoting the SDK as the "one stop shop" for devs. My gut feeling says that only Linux repo maintainers and OpenCL implementers would want to build this repo directly. Other users will clone the SDK and build using -D `OPENCL_SDK_BUILD_SAMPLES=OFF` if looking for the bare minimum.

Reply:
@jenatali thank you for the tip on `FetchContent`. Allow me to add: you're living the life, being able to use CMake 3.11. :) Here we're still constrained to 3.1. It's a self-imposed restriction, but it is what it is. I did have to increase the minimum for the SDK to 3.10, but [only when building OpenGL samples](https://github.com/StreamHPC/OpenCL-SDK/blob/develop_stream/CMakeLists.txt#L46-L47) because SFML simply wouldn't budge below that. I can't take on the burden of backporting the scripts of multimedia windowing libraries. They are massive and messy. Nor do I want to reimplement any parts of the libs to omit depending on them.

Reply:
I've been playing around with this and updating some or my libraries over the weekend and I really think we need to re-add a method of specifying OpenCL headers that does not require installation or we're going to break a lot of automation and end up with grumpy users.

What if we do something like this:

1. Check if the `OPENCL_ICD_LOADER_HEADERS_DIR` exists and contains OpenCL headers.  If it does, use them.  This will keep existing toolchains functioning.  It also means that the OpenCL ICD loader can be built with just one invocation of CMake, if desired.
2. If `OPENCL_ICD_LOADER_HEADERS_DIR` doesn't exist and the `OpenCL::Headers` target doesn't exist, try to `find_package(OpenCLHeaders)`.  This enables the new fully CMake toolchain.
3. If desired, we can investigate a `FetchContent` solution for newer versions of CMake if the `OpenCL::Headers` target _still_ doesn't exist, but (1) and (2) should cover most usages.

Does this sound doable?

Reply:
Re-added hardcoded header paths variable.